code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -73,36 +113,173 @@ public void computeInstanceActivateReply(Event event) {
+                onHost.remove(reported.getValue().getExternalId());
+                inRancher.remove(reported.getKey());
+                reported.getValue().setInstance(ki);
+                determineSyncAction(ki, reported.getValue(), needsSynced, syncActions, checkOnly);
+            }
+        }
+
+        if (!onHost.isEmpty() || !inRancher.isEmpty()) {
+            Map<String, Instance> knownByExternalId = new HashMap<String, Instance>();
+            for (Instance ki : knownInstances.values()) {
+                if (StringUtils.isNotEmpty(ki.getExternalId()))
+                    knownByExternalId.put(ki.getExternalId(), ki);
+            }
+
+            for (Map.Entry<String, ReportedInstance> reported : reportedInstances.byExternalId.entrySet()) {
+                Instance ki = knownByExternalId.get(reported.getKey());
+                if (ki != null) {
+                    onHost.remove(reported.getKey());
+                    inRancher.remove(ki.getUuid());
+                    reported.getValue().setInstance(ki);
+                    determineSyncAction(ki, reported.getValue(), needsSynced, syncActions, checkOnly);
+                }
+            }
+        }
+
+        // Anything left in onHost is on the host, but not in rancher.
+        for (Map.Entry<String, ReportedInstance> create : onHost.entrySet()) {
+            ReportedInstance ri = create.getValue();
+            if (StringUtils.isNotEmpty(ri.getLabels().get(LABEL_RANCHER_SYSTEM_CONTAINER))) {
+                // Unknown system container. Force stop.
+                if (!STATE_STOPPED.equals(ri.getState())) {
+                    addSyncAction(needsSynced, syncActions, ri, EVENT_INSTANCE_FORCE_STOP, checkOnly);
                 }
+                continue;
             }
+            addSyncAction(needsSynced, syncActions, ri, EVENT_CREATE, checkOnly);
         }
 
-        return false;
+        // Anything left in inRancher is in rancher, but not on the host.
+        for (Instance ki : inRancher.values()) {
+            if (objectMetaDataManager.isTransitioningState(Instance.class, ki.getState()) || STATE_STOPPED.equals(ki.getState()) || ki.getRemoved() != null)
+                continue;
+
+            if (StringUtils.isNotEmpty(ki.getSystemContainer()) || StringUtils.isEmpty(ki.getExternalId())) {
+                // System container or not enough info to perform no-op containerEvent. Schedule potential restart.
+                // This is the one place we can't use addSyncAction, since we don't have (and can't construct) a ReportedInstance.
+                if (checkOnly) {
+                    throw new ContainersOutOfSync();
+                }
+                needsHaRestart.add(ki.getUuid());
+            } else {
+                ReportedInstance ri = new ReportedInstance();
+                ri.setExternalId(ki.getExternalId());
+                ri.setUuid(ki.getUuid());
+                addSyncAction(needsSynced, syncActions, ri, EVENT_STOP, checkOnly);
```

Comment:
@alena1108 I added `hasInstanceTriggeredStopConfigured()` to the check. So, we'll do the HA_RESTART either if it is a system container or it has that property set (to something other than stop, which they all have by default).

I did this because we never want ContainEvents to manage system containers. This is because ContainerEvents assume that the host is the source of truth for the intended state of the container, but for system containers, Rancher is _always_ the source of truth.